### PR TITLE
Migrating from ASP.NET MVC5 to MVC6, Part 1

### PR DESCRIPTION
**Pull request by ardalis**
This is the first article. It was getting long, and there are features of the new MVC 6 starter template that aren't feature complete yet, so I'm breaking it up into a series. In fact, as long as this article is, it might be better broken up already (shifting the configuration of bundling and minification to its own separate article).

### CODE REVIEW

**Comment by ardalis** [comment]
Add migrating config article as well. Skeleton of auth/identity article can be ignored for now.

File: docs/github-test.rst [hunk 1]
```diff
@@ -0,0 +1,258 @@
+Testing to See if GitHub Breaks RST Files When Editing
+======================================================
+By `Steve Smith`_ | Originally Published: 1 June 2015 
+
+.. _`Steve Smith`: Author_
+
+Migrating from ASP.NET MVC 5 to ASP.NET 5 and MVC 6 requires a few steps to complete, since ASP.NET 5 introduces a number of new concepts. In this article you will learn how to migrate from the ASP.NET MVC 5 Starter Web Project to ASP.NET MVC 6, including initial setup, basic controllers and views, and static content and client side dependencies.
```
Comment (danroth27) [inline docs/github-test.rst]: > migrate from the ASP.NET MVC 5 Starter Web Project 

migrate from the ASP.NET MVC 5 default project template

Comment (danroth27) [inline docs/github-test.rst]: > and static content

~~and~~ static content

File: docs/github-test.rst [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	- Create the Destination Solution
+	- Migrate Basic Controllers, Views, and Static Content
+	- Configure Bundling
+
+You can download the finished source from the project created in this article HERE **(TODO)**.
+
+Create the Initial Project
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For the purposes of this article, we will be starting from the default ASP.NET MVC 5 starter web project, which you can create in Visual Studio 2015 by adding a new web project, and choosing MVC 5.
```
Comment (danroth27) [inline docs/github-test.rst]: project~~,~~ and

File: docs/github-test.rst [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+.. image:: _static/new-project.png
+
+.. image:: _static/new-project-select-mvc-template.png
+
+If you prefer, you can `download the MVC5Project used in this article </_static/mvc5project.zip>`_. **(TODO)**
```
Comment (danroth27) [inline docs/github-test.rst]: > MVC5Project 

MVC 5 Project

File: docs/github-test.rst [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+This sample web project will demonstrate how to migrate an MVC 5 web project that includes controllers, views, and ASP.NET Identity models, as well as startup and configuration logic common to many MVC 5 projects.
+
+Create the Destination Solution
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We will begin our migration by creating a new, empty ASP.NET 6 solution. Create a new project in Visual Studio 2015, choose an ASP.NET Web Application, and then choose select the ASP.NET 5 Empty template.
```
Comment (danroth27) [inline docs/github-test.rst]: > ASP.NET 6

ASP.NET 5

Comment (danroth27) [inline docs/github-test.rst]: > choose select

choose ~~select~~

File: docs/github-test.rst [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+.. image:: _static/new-project-mvc6.png
+
+.. image:: _static/new-project-select-empty-aspnet5-template.png
+
+This migration will start from an empty template. If you're already familiar with ASP.NET 6 and its starter templates and there are features in a starter template you would like to take advantage of, you may wish to start from another template. The next step is to configure the site to use MVC. This requires changes to the project.json file and startup.cs file. First, open project.json and add "Microsoft.AspNet.Mvc" to the "dependencies" property:
```
Comment (danroth27) [inline docs/github-test.rst]: > ASP.NET 6

ASP.NET 5

Comment (danroth27) [inline docs/github-test.rst]: > startup.cs

Startup.cs

File: docs/github-test.rst [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+.. code-block:: c#
```
Comment (danroth27) [inline docs/github-test.rst]: json

File: docs/github-test.rst [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+	"dependencies": {
```
Comment (danroth27) [inline docs/github-test.rst]: beta3 => beta4

File: docs/github-test.rst [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+	public void ConfigureServices(IServiceCollection services)
+	{
+		services.AddMvc();
+	}
+
+	public void Configure(IApplicationBuilder app)
+	{
+		app.UseMvc(routes =>
+		{
+			routes.MapRoute(
```
Comment (danroth27) [inline docs/github-test.rst]:
```csharp
routes.MapRoute(
    name: "default",
    template: "{controller=Home}/{action=Index}/{id?}");

File: docs/github-test.rst [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+				name: "default",
+				template: "{controller}/{action}/{id?}",
+				defaults: new { controller = "Home", action = "Index" });
+		});
+	}
+
+At this point we are ready to create a simple Controller and View. Add a Controllers folder and a Views folder to the project. Add an MVC Controller called HomeController.cs class to the Controllers folder and a Home folder in the Views folder. Finally, add an Index.cshtml MVC View Page to the Views/Home folder. The project structure should be as shown:
```
Comment (danroth27) [inline docs/github-test.rst]: > a simple Controller and View

Use consistent casing for Controller and View throughout. The folder names should always be capitalized.

File: docs/github-test.rst [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+Copy each of the methods from the source HomeController to the HomeController we added to the project in the previous section. Note that in MVC 5, actions typically returned ActionResult, but in MVC 6 this has changed to IActionResult (though it will still compile if you leave it as ActionResult).
+
+Next, create new MVC View Pages in the Views/Home folder for About and Contact. Copy the contents of the corresponding views in the old project to these new views, as well as Index.cshtml. At this point you should once again be able to run the new application, and although the styles are not yet in place, you should see the correct content on the home page as well as /home/about and /home/contact (contact is shown here):
+
+.. image:: _static/contact-page.png
+
+In MVC5 and previous versions of ASP.NET, static content was hosted from the root of the web project, and was intermixed with server-side files. In MVC6, all static content files are hosted from the /wwwroot folder, so we will need to adjust where we are storing our static content files. For instance, we can copy the favicon.ico file from the root of the original project to the /wwwroot folder in the new project.
```
Comment (danroth27) [inline docs/github-test.rst]: MVC5 => MVC 5

Comment (danroth27) [inline docs/github-test.rst]: MVC6 => MVC 6

File: docs/github-test.rst [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+The MVC5 project uses Bootstrap for its styling, with files stored in /Content and /Scripts and referenced in /Views/Shared/_Layout.cshtml. We could simply copy the bootstrap.js and bootstrap.css files from the old project to the /wwwroot folder in the new project, but there are better ways to handle these kinds of client-side library dependencies in ASP.NET 5.
```
Comment (danroth27) [inline docs/github-test.rst]: MVC5 => MVC 5

File: docs/github-test.rst [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+
+In our new project, we'll add support for Bootstrap (and other client-side libraries), but we'll do so using the new support for client-side build tooling using bower and grunt. First, add a new Bower JSON Configuration File to the project root, called bower.json. In its "dependencies" property, add bootstrap, jquery, jquery-validation, and jquery-validation-unobtrusive. Add new properties for these items to the "exportsOverride" property as well, so that the complete bower.json file looks like this:
```
Comment (danroth27) [inline docs/github-test.rst]: Consistently capitalize Bower and Grunt

Comment (danroth27) [inline docs/github-test.rst]: We recently decided to go with Gulp instead of Grunt in our templates.

File: docs/github-test.rst [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+	{
+		"name": "NewMvc6Project",
+		"private": true,
+		"dependencies": {
+			"bootstrap": "3.0.0",
+			"jquery": "1.10.2",
+			"jquery-validation": "1.11.1",
+			"jquery-validation-unobtrusive": "3.2.2"
+		},
+		"exportsOverride": {
```
Comment (danroth27) [inline docs/github-test.rst]: @DamianEdwards @madskristensen Are these exportsOverrides really needed?

File: docs/github-test.rst [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+				"": "jquery.validate.unobtrusive.{js,min.js}"
+			}
+		}
+	}
+
+Bower will automatically download the specified dependencies, but for now the files are not yet in the wwwroot folder, and so cannot be requested by the browser:
+
+.. image:: _static/project-structure-bower.png
+
+Next, we will configure Grunt to process these files and place them where we want them in the wwwroot folder. First, we need to make sure Grunt is installed locally for the project. This is accomplished using NPM, which is similar to Bower but requires a different configuration file, "package.json". Add a new NPM configuration file to the root of the project, called package.json. Add the "grunt" and "grunt-bower-task" items to the devDependencies property, as shown (you should get Intellisense as you type their names):
```
Comment (danroth27) [inline docs/github-test.rst]: Need  to update this content to use Gulp as reflected in our latest templates.

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+.. image:: _static/new-project-select-mvc-template.png
+
+If you prefer, you can `download the MVC5Project used in this article </_static/mvc5project.zip>`_. **(TODO)**
+
+This sample web project will demonstrate how to migrate an MVC 5 web project that includes controllers, views, and ASP.NET Identity models, as well as startup and configuration logic common to many MVC 5 projects.
+
+Create the Destination Solution
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We will begin our migration by creating a new, empty ASP.NET 6 solution. Create a new project in Visual Studio 2015, choose an ASP.NET Web Application, and then choose select the ASP.NET 5 Empty template.
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: ASP.NET 6 => ASP.NET 5

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+
+.. image:: _static/new-project-mvc6.png
+
+.. image:: _static/new-project-select-empty-aspnet5-template.png
+
+This migration will start from an empty template. If you're already familiar with ASP.NET 6 and its starter templates and there are features in a starter template you would like to take advantage of, you may wish to start from another template. The next step is to configure the site to use MVC. This requires changes to the project.json file and startup.cs file. First, open project.json and add "Microsoft.AspNet.Mvc" to the "dependencies" property:
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: ASP.NET 6 => ASP.NET 5

Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: startup.cs => Startup.cs

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+
+.. code-block:: c#
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: c# => json

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+
+	"dependencies": {
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: beta3 => beta4

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+	public void ConfigureServices(IServiceCollection services)
+	{
+		services.AddMvc();
+	}
+
+	public void Configure(IApplicationBuilder app)
+	{
+		app.UseMvc(routes =>
+		{
+			routes.MapRoute(
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]:
```csharp
routes.MapRoute(
    name: "default",
    template: "{controller=Home}/{action=Index}/{id?}");
```

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+				name: "default",
+				template: "{controller}/{action}/{id?}",
+				defaults: new { controller = "Home", action = "Index" });
+		});
+	}
+
+At this point we are ready to create a simple Controller and View. Add a Controllers folder and a Views folder to the project. Add an MVC Controller called HomeController.cs class to the Controllers folder and a Home folder in the Views folder. Finally, add an Index.cshtml MVC View Page to the Views/Home folder. The project structure should be as shown:
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: Use consistent casing for controller and view. The folder names should always be capitalized.

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+
+Copy each of the methods from the source HomeController to the HomeController we added to the project in the previous section. Note that in MVC 5, actions typically returned ActionResult, but in MVC 6 this has changed to IActionResult (though it will still compile if you leave it as ActionResult).
+
+Next, create new MVC View Pages in the Views/Home folder for About and Contact. Copy the contents of the corresponding views in the old project to these new views, as well as Index.cshtml. At this point you should once again be able to run the new application, and although the styles are not yet in place, you should see the correct content on the home page as well as /home/about and /home/contact (contact is shown here):
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: Need to cover what to do with web.config in Views folder. Convert to _GlobalImport.cshtml with `@using` statements.

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+	{
+		"name": "NewMvc6Project",
+		"private": true,
+		"dependencies": {
+			"bootstrap": "3.0.0",
+			"jquery": "1.10.2",
+			"jquery-validation": "1.11.1",
+			"jquery-validation-unobtrusive": "3.2.2"
+		},
+		"exportsOverride": {
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: @DamianEdwards @madskristensen Is this exportsOverride section needed?

File: docs/migrating/migratingfrommvc5/migratingfrommvc5.rst [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+				"": "jquery.validate.unobtrusive.{js,min.js}"
+			}
+		}
+	}
+
+Bower will automatically download the specified dependencies, but for now the files are not yet in the wwwroot folder, and so cannot be requested by the browser:
+
+.. image:: _static/project-structure-bower.png
+
+Next, we will configure Grunt to process these files and place them where we want them in the wwwroot folder. First, we need to make sure Grunt is installed locally for the project. This is accomplished using NPM, which is similar to Bower but requires a different configuration file, "package.json". Add a new NPM configuration file to the root of the project, called package.json. Add the "grunt" and "grunt-bower-task" items to the devDependencies property, as shown (you should get Intellisense as you type their names):
```
Comment (danroth27) [inline docs/migrating/migratingfrommvc5/migratingfrommvc5.rst]: We have switched from Grunt to Gulp, so unfortunately the Grunt content will need to be updated to be based on Gulp.

**Comment by danroth27** [comment]
Let's not use zip files. I prefer to see the sample projects checked in in a samples directory that can be built and referenced.

**Comment by danroth27** [comment]
This is a good start for an MVC 5 migration doc, but I think there is a bunch of stuff that is different in MVC 6 that we still need to cover:
- Areas
- Global filters
- MVC options configuration in general
- DI
- Attribute routing

**Comment by ardalis** [comment]
Thanks for the feedback - my plan is to have additional articles to cover the additional MVC 6 features, while keeping this one a bit simpler and shorter.

**Pull request merged**